### ISSUE TITLE
Third order does not work with official stable OS X package (4.5.1)

### ISSUE THREAD

**Issue by stepan-a** [DynareTeam/dynare#1492]
I obtain the following error message:
```matlab
Invalid MEX-file '/Applications/Dynare/4.5.1/matlab/../mex/matlab/osx/k_order_perturbation.mexmaci64': Missing dependent shared libraries:
'@rpath/libhdf5.6.dylib' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'

Missing symbol '__DefaultRuneLocale' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '__Unwind_Resume' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___bzero' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___cxa_atexit' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___maskrune' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___memcpy_chk' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___powidf2' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___stack_chk_fail' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___stack_chk_guard' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___stderrp' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___stdoutp' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '___vsnprintf_chk' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_abort' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_asctime_r' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_calloc' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_clock' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_ctime' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_dlclose' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_dlerror' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_dlopen' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_dlsym' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_drand48' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_exit' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_fclose' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_feof' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_fflush' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_floor' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_fopen' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_fprintf' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_fread' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_free' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_fseek' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_ftell' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_fwrite' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_getloadavg' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_getrusage' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_gettimeofday' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_localtime_r' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_log' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_malloc' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_memcpy' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_memmove' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_memset' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_mktemp' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pow' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_printf' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_attr_destroy' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_attr_init' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_attr_setdetachstate' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_cond_broadcast' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_cond_destroy' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_cond_init' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_cond_wait' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_create' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_exit' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_join' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_mutex_init' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_mutex_lock' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_pthread_mutex_unlock' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_putchar' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_puts' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_remove' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_rename' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_round' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_sprintf' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_srand48' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_strcat' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_strcmp' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_strcpy' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_strdup' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_strlen' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_strncpy' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_sysconf' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_time' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_uname' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'
Missing symbol '_vasprintf' required by '/Applications/Dynare/4.5.1/mex/matlab/osx/k_order_perturbation.mexmaci64'.

Error in k_order_pert (line 57)
        [err, g_0, g_1, g_2, g_3, derivs] = k_order_perturbation(dr, ...

Error in stochastic_solvers (line 96)
        [dr,info] = k_order_pert(dr,M_,options_);

Error in resol (line 144)
    [dr,info] = stochastic_solvers(dr,check_flag,M,options,oo);

Error in stoch_simul (line 84)
    [oo_.dr,info,M_,options_,oo_] = resol(0,M_,options_,oo_);
```
Note however that if Dynare is installed through Homebrew, there is no such problems with `k_order_perturbation` Mex file. @houtanb Is it possible to have access to the code generating the package?

**Comment by houtanb** [comment]
@stepan-a I'll work on this today. For future reference, the code is here: https://github.com/houtanb/snapshot-macOS

**Comment by stepan-a** [comment]
Thanks for the link.

**Comment by houtanb** [comment]
This has been fixed. `hdf5` was being linked dynamically. This will be fixed in the snapshot the next time it is created.

**Comment by stepan-a** [comment]
@houtanb I understood the problem, but can you explain where you fixed it? I do not see a patch on the repo you were referring to above...

**Comment by stepan-a** [comment]
This needs also to be in the next stable release (4.5.2). I would like to do it next week.

**Comment by houtanb** [comment]
@stepan-a it's on the computer that compiles it. I just needed to copy the static `hd5` libraries to `/usr/local/static/lib`

**Comment by stepan-a** [comment]
What I still do not understand is that I have `hdf5` through Homebrew on my laptop (that's why the version shipped with Homebrew or compiled works)...

**Comment by houtanb** [comment]
@stepan-a I don't understand your comment. Can you rephrase it?

**Comment by stepan-a** [comment]
I have `hdf5` library installed on my laptop (with Homebrew). I do not understand why the `k_order_perturbation` mex distributed in the official package does not find/use it.

**Comment by houtanb** [comment]
`otool -L k_order_perturbation.mexmaci64` on the bad k_order mex file yields the line:
```
@rpath/libhdf5.6.dylib (compatibility version 7.0.0, current version 7.5.0)
```
For me, `ls /usr/local/lib/libhdf5.6.dylib` tells me it is not found. This is because the version of hdf5 on the computer that is building the snapshot and release is older than the current version (I have not updated it)

**Comment by stepan-a** [comment]
Thanks, I understand. Shoulldn't we ensure that we use the same libraries with the packaged and homebrew shipped versions of Dynare?

**Comment by houtanb** [comment]
I don't think there is anything that imposes that. The homebrew version of Dynare is supposed to be used for use on Octave whereas the version we release is for Matlab. There are already enough differences between Octave and Matlab libraries that these other libraries being the same won't matter.

Furthermore, compiling the static libraries we need before every release would add quite some time to the release schedule. That said, I'm not against updating them more regularly.
